lib/puppet/indirector/indirection.rb

```diff
@@ -101,10 +105,28 @@ def initialize(model, name, doc: nil, indirected_class: nil, cache_class: nil, t
     @indirected_class = indirected_class
     self.extend(extend) if extend
 
-    # These setters depend on the indirection already being installed so they have to be at the end
-    self.cache_class = cache_class if cache_class
-    self.terminus_class = terminus_class if terminus_class
-    self.terminus_setting = terminus_setting if terminus_setting
+    # Setting these depend on the indirection already being installed so they have to be at the end
+    validate_terminus_class(cache_class) if !cache_class.nil?
+    validate_terminus_class(terminus_class) if !terminus_class.nil?
+    @cache_class = Puppet::ThreadLocal.new(cache_class)
+    @terminus_class = Puppet::ThreadLocal.new(terminus_class)
+    @terminus_setting = Puppet::ThreadLocal.new(terminus_setting)
+  end
+
+  # Use this to set indirector settings globally across threads.
+  def set_global_setting(setting, value)
+    case setting
+    when :cache_class
+      validate_terminus_class(value) if !value.nil?
```

Comment:
nit: I don't think `false` is ever a valid value here?
If so, we could change this to `if value` rather than if-not-value-nil.

Reply:
I copied these nil checks from the previous/local setters, but since this is an entirely new method, maybe we don't need to be as defensive here.